Document kernel version requirements for Linux (#47363)
kubernetes/website: content/en/docs/reference/node/kernel-version-requirements.md

---
content_type: "reference"

> title: OS Kernel Version Requirements

Reviewer: ```suggestion
title: Linux Kernel Version Requirements
```

weight: 10
---

> Many features rely on specific kernel functionalities and have minimum kernel version requirements.

Reviewer: Please add the third party software disclaimer using a [shortcode](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#third-party-content-marker); things mentioned here are `runc` and Linux.

> ## Pod [sysctls](/docs/tasks/administer-cluster/sysctl-cluster/#safe-and-unsafe-sysctls)
> 
> `sysctl` is a Linux-specific command-line tool used to configure various kernel parameters
> and it is not available on non-Linux operating systems.

Reviewer: Avoid hyperlinks within headings.

```suggestion
## Pod sysctls {#requirements-pod-sysctl}

On Linux, the `sysctl()` system call configures kernel parameters at run time. There is a command
line tool named `sysctl` that you can use to configure these parameters, and many are exposed via
the `proc` filesystem.

Some sysctls are only available if you have a modern enough kernel.
```

> The following sysctls are supported in the safe set, which has a minimal kernel version requirement:

Reviewer: ```suggestion
The following sysctls are supported in the [safe set](/docs/tasks/administer-cluster/sysctl-cluster/#safe-and-unsafe-sysctls)
which has a minimal kernel version requirement:
```

- `net.ipv4.ip_local_reserved_ports` (since Kubernetes 1.27, needs kernel 3.16+);
- `net.ipv4.tcp_keepalive_time` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_fin_timeout` (since Kubernetes 1.29, needs kernel 4.6+);
- `net.ipv4.tcp_keepalive_intvl` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_keepalive_probes` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_syncookies` (namespaced since kernel 4.6+).
- `net.ipv4.vs.conn_reuse_mode` (used in `ipvs` proxy mode, needs kernel 4.1+);

### kube proxy `nftables` proxy mode

[`nftables` proxy mode](/docs/reference/networking/virtual-ips/#proxy-mode-nftables) is introduced in v1.29.
The (alpha) nftables mode of kube-proxy now requires version 1.0.1 or later
of the nft command-line, and kernel 5.13 or later. (For testing/development
purposes, you can use older kernels, as far back as 5.4, if you set the
`nftables.skipKernelVersionCheck` option in the kube-proxy config, but this is not
recommended in production since it may cause problems with other nftables

> users on the system.

Reviewer: This will become stale with the v1.31 release. Can we write it to be more neutral as to whether nftables support is alpha / beta / stable?

Author: Removed the alpha description here and only mentioned that the kernel version requiement.

> ## [Cgroup v2](/docs/concepts/architecture/cgroups/#requirements) requirements

Reviewer: Avoid hyperlinks within headings.
```suggestion
## Version 2 control groups
```

> Cgroup v2 requires that Linux Kernel version is 5.8 or later.

Reviewer: ```suggestion
cgroup v1 support is being deprecated / moved to a less maintained status with v1.31, so very soon _Kubernetes_ Linux nodes will at least recommend this kernel version.
```

In [Linux 5.8](https://github.com/torvalds/linux/commit/4a7e89c5ec0238017a757131eb9ab8dc111f961c), the system-level `cpu.stat` file was added to the root cgroup for convenience.

In [runc document](https://github.com/containerd/cgroups/blob/0c03de4a3d82a5f02f455ccc8174cb0dc9c2a532/cgroup2/manager.go#L411-L430), Kernel older than 5.2 is not recommended due to lack of freezer.

> ## Others kernel requirements

Reviewer: ```suggestion
## Other kernel requirements {#requirements-other}
```

Some features may depend on new kernel functionalities and have specific kernel requirements:

1. [Recursive read only mount](/docs/concepts/storage/volumes/#recursive-read-only-mounts): This is implemented by applying the `MOUNT_ATTR_RDONLY` attribute with the `AT_RECURSIVE` flag using `mount_setattr`(2) added in Linux kernel v5.12.
2. Pod user namespace support requires minimal kernel version 6.5+, according to [KEP-127](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/127-user-namespaces/README.md).
3. For [node system swap](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md), tmpfs noswap is not supported until kernel 6.3.

> ## Longterm release kernel versions

Reviewer: ```suggestion
## Linux kernel long term maintenance
```

Active kernel releases can be found in [kernel.org](https://www.kernel.org/category/releases.html).

> There are usually several "longterm maintenance" kernel releases provided for the purposes of backporting bugfixes for older kernel trees. Only important bugfixes are applied to such kernels and they don't usually see very frequent releases, especially for older trees. Longterm release kernels like 4.19, 5.15.

Reviewer: ```suggestion
There are usually several _long term maintenance_ kernel releases provided for the purposes of backporting
bugfixes for older kernel trees. Only important bugfixes are applied to such kernels and they don't
usually see very frequent releases, especially for older trees.

See the Linux kernel website for the [list of releases](https://www.kernel.org/category/releases.html)
in the _Longterm_ category.
```

---
content_type: "reference"
title: Linux Kernel Version Requirements
weight: 10
---

{{% thirdparty-content %}}

Many features rely on specific kernel functionalities and have minimum kernel version requirements.

> ## Pod sysctls {#requirements-pod-sysctl}

Reviewer 2: Would be great to have comment in this page pointing back to the constants in kubelet for each of the mentioned limitations. This way maintaining this page will be easier long term

```
<!--
Code: link to the constant in k/k
-->
```

Author: https://github.com/kubernetes/website/pull/47363/commits/19a4a014b74ba5e751239840f32d52be087a4cce Added some code reference.

which has a minimal kernel version requirement:

- `net.ipv4.ip_local_reserved_ports` (since Kubernetes 1.27, needs kernel 3.16+);
- `net.ipv4.tcp_keepalive_time` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_fin_timeout` (since Kubernetes 1.29, needs kernel 4.6+);
- `net.ipv4.tcp_keepalive_intvl` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_keepalive_probes` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_syncookies` (namespaced since kernel 4.6+).
- `net.ipv4.vs.conn_reuse_mode` (used in `ipvs` proxy mode, needs kernel 4.1+);

### kube proxy `nftables` proxy mode

The [`nftables` mode](/docs/reference/networking/virtual-ips/#proxy-mode-nftables) of kube-proxy now requires version 1.0.1 or later
of the nft command-line, and kernel 5.13 or later. (For testing/development
purposes, you can use older kernels, as far back as 5.4, if you set the
`nftables.skipKernelVersionCheck` option in the kube-proxy config, but this is not
recommended in production since it may cause problems with other nftables
users on the system.

## Version 2 control groups

> cgroup v1 support is being deprecated / moved to a less maintained status with v1.31, so very soon _Kubernetes_ Linux nodes will at least recommend this kernel version.

Reviewer 2: ```suggestion
cgroup v1 support is in maintained mode stating v1.31, and cgroup v2 is a recommended. So very soon _Kubernetes_ Linux nodes will at least recommend this kernel version.
```

the `proc` filesystem.

Some sysctls are only available if you have a modern enough kernel.

The following sysctls are supported in the [safe set](/docs/tasks/administer-cluster/sysctl-cluster/#safe-and-unsafe-sysctls)
which has a minimal kernel version requirement:

<!--
Code: https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/pkg/util/kernel/constants.go#L19-L45
-->
- `net.ipv4.ip_local_reserved_ports` (since Kubernetes 1.27, needs kernel 3.16+);
- `net.ipv4.tcp_keepalive_time` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_fin_timeout` (since Kubernetes 1.29, needs kernel 4.6+);
- `net.ipv4.tcp_keepalive_intvl` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_keepalive_probes` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_syncookies` (namespaced since kernel 4.6+).
- `net.ipv4.vs.conn_reuse_mode` (used in `ipvs` proxy mode, needs kernel 4.1+);

### kube proxy `nftables` proxy mode

<!--
Code: https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/pkg/util/kernel/constants.go#L53-L56
-->

> The [`nftables` mode](/docs/reference/networking/virtual-ips/#proxy-mode-nftables) of kube-proxy now requires version 1.0.1 or later

Reviewer: ```suggestion
For Kubernetes {{< skew currentVersion >}}, the
[`nftables` mode](/docs/reference/networking/virtual-ips/#proxy-mode-nftables) of kube-proxy requires
version 1.0.1 or later
```

In [runc document](https://github.com/containerd/cgroups/blob/0c03de4a3d82a5f02f455ccc8174cb0dc9c2a532/cgroup2/manager.go#L411-L430), Kernel older than 5.2 is not recommended due to lack of freezer.

## Other kernel requirements {#requirements-other}

Some features may depend on new kernel functionalities and have specific kernel requirements:

<!--
Code(recursive read only mount): https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L1605-L1609
Code(user namespace and swap): https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/pkg/util/kernel/constants.go#L47-L51
-->
1. [Recursive read only mount](/docs/concepts/storage/volumes/#recursive-read-only-mounts): This is implemented by applying the `MOUNT_ATTR_RDONLY` attribute with the `AT_RECURSIVE` flag using `mount_setattr`(2) added in Linux kernel v5.12.
2. Pod user namespace support requires minimal kernel version 6.5+, according to [KEP-127](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/127-user-namespaces/README.md).

> 3. For [node system swap](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md), tmpfs noswap is not supported until kernel 6.3.

Reviewer: ```suggestion
3. For [node system swap](/docs/concepts/architecture/nodes/#swap-memory), tmpfs set to `noswap` is not supported until kernel 6.3.
  This has security implications; read the associated documentation to learn more.
```

Author: Updated.

---
content_type: "reference"
title: Linux Kernel Version Requirements
weight: 10
---

{{% thirdparty-content %}}

Many features rely on specific kernel functionalities and have minimum kernel version requirements.
However, relying solely on kernel version numbers may not be sufficient
for certain operating system distributions,

> as kernel vendors such as RHEL, Ubuntu and SUSE often backport features to older kernels.

Reviewer: How about:
```suggestion
as maintainers for distributions such as RHEL, Ubuntu and SUSE often backport selected features
to older kernel releases (retaining the older kernel version).
```

## Pod sysctls {#requirements-pod-sysctl}

On Linux, the `sysctl()` system call configures kernel parameters at run time. There is a command

> line tool named `sysctl` that you can use to configure these parameters, and many are exposed via

Author: This is a general description.
There may be too many things to explain here.

the `proc` filesystem.

> Some sysctls are only available if you have a modern enough kernel.

Reviewer: @tengqm, without the “enough” this would not look like idiomatic English.

The following sysctls are supported in the [safe set](/docs/tasks/administer-cluster/sysctl-cluster/#safe-and-unsafe-sysctls)

> which has a minimal kernel version requirement:

Author: For `following sysctls`.

I changed to 
```
The following sysctls have a minimal kernel version requirement,
and are supported in the [safe set](/docs/tasks/administer-cluster/sysctl-cluster/#safe-and-unsafe-sysctls):
```

<!--
Code: https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/pkg/util/kernel/constants.go#L19-L45
-->
- `net.ipv4.ip_local_reserved_ports` (since Kubernetes 1.27, needs kernel 3.16+);
- `net.ipv4.tcp_keepalive_time` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_fin_timeout` (since Kubernetes 1.29, needs kernel 4.6+);
- `net.ipv4.tcp_keepalive_intvl` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_keepalive_probes` (since Kubernetes 1.29, needs kernel 4.5+);
- `net.ipv4.tcp_syncookies` (namespaced since kernel 4.6+).
- `net.ipv4.vs.conn_reuse_mode` (used in `ipvs` proxy mode, needs kernel 4.1+);

### kube proxy `nftables` proxy mode

<!--
Code: https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/pkg/util/kernel/constants.go#L53-L56
-->
For Kubernetes {{< skew currentVersion >}}, the
[`nftables` mode](/docs/reference/networking/virtual-ips/#proxy-mode-nftables) of kube-proxy requires
version 1.0.1 or later

> of the nft command-line tool, as well as kernel 5.13 or later. (For testing/development

Reviewer: The interface with nftables is deliberately via `nft`, because the netlink interface is poorly documented with no nice-to-use Go client code.

Maybe a future release of `kube-proxy` will switch to using netlink here.

> of the nft command-line tool, as well as kernel 5.13 or later. (For testing/development

Author: /cc @danwinship 
to confirm

recommended in production since it may cause problems with other nftables
users on the system.

## Version 2 control groups

Kubernetes cgroup v1 support is in maintained mode starting from Kubernetes v1.31; using cgroup v2
is recommended.
In [Linux 5.8](https://github.com/torvalds/linux/commit/4a7e89c5ec0238017a757131eb9ab8dc111f961c), the system-level `cpu.stat` file was added to the root cgroup for convenience.

In [runc document](https://github.com/containerd/cgroups/blob/0c03de4a3d82a5f02f455ccc8174cb0dc9c2a532/cgroup2/manager.go#L411-L430), Kernel older than 5.2 is not recommended due to lack of freezer.

## Other kernel requirements {#requirements-other}

Some features may depend on new kernel functionalities and have specific kernel requirements:

<!--
Code(recursive read only mount): https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto#L1605-L1609
Code(user namespace and swap): https://github.com/kubernetes/kubernetes/blob/00236ae0d73d2455a2470469ed1005674f8ed61f/pkg/util/kernel/constants.go#L47-L51
-->

> 1. [Recursive read only mount](/docs/concepts/storage/volumes/#recursive-read-only-mounts): This is implemented by applying the `MOUNT_ATTR_RDONLY` attribute with the `AT_RECURSIVE` flag using `mount_setattr`(2) added in Linux kernel v5.12.

Reviewer 3: ```suggestion
1. [Recursive read only mount](/docs/concepts/storage/volumes/#recursive-read-only-mounts):
   This is implemented by applying the `MOUNT_ATTR_RDONLY` attribute with the `AT_RECURSIVE` flag
   using `mount_setattr`(2) added in Linux kernel v5.12.
```

> 2. Pod user namespace support requires minimal kernel version 6.5+, according to [KEP-127](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/127-user-namespaces/README.md).

Reviewer 3: ```suggestion
2. Pod user namespace support requires minimal kernel version 6.5+, according to
   [KEP-127](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/127-user-namespaces/README.md).
```

> 3. For [node system swap](/docs/concepts/architecture/nodes/#swap-memory), tmpfs set to `noswap` is not supported until kernel 6.3.

Reviewer 3: ```suggestion
3. For [node system swap](/docs/concepts/architecture/nodes/#swap-memory), tmpfs set to `noswap`
   is not supported until kernel 6.3.
```

  This has security implications; read the associated documentation to learn more.

## Linux kernel long term maintenance

Active kernel releases can be found in [kernel.org](https://www.kernel.org/category/releases.html).

There are usually several _long term maintenance_ kernel releases provided for the purposes of backporting

> bugfixes for older kernel trees. Only important bugfixes are applied to such kernels and they don't

Reviewer 3: ```suggestion
bug fixes for older kernel trees. Only important bug fixes are applied to such kernels and they don't
```

---
content_type: "reference"
title: Linux Kernel Version Requirements
weight: 10
---

{{% thirdparty-content %}}

Many features rely on specific kernel functionalities and have minimum kernel version requirements.
However, relying solely on kernel version numbers may not be sufficient
for certain operating system distributions,

> as maintainers for distributions such as RHEL, Ubuntu and SUSE often backport selected features

Reviewer: @pacoxu we should list these vendors in alphabetical order. Willing to send in a follow up PR?

to older kernel releases (retaining the older kernel version).

> ## Pod sysctls {#requirements-pod-sysctl}

Reviewer 3: ```suggestion
## Pod sysctls
```

Author: Updated.
